File: 1013. Partition Array Into Three Parts With Equal Sum/solution.cpp
class Solution {
public:
  bool canThreePartsEqualSum(vector<int>& arr) {
    int left = arr[0];
    int right = arr[arr.size() - 1];
    int total = 0;
    for (int index = 0; index < arr.size(); index++){
      int num = arr[index];
      total += num;
    }

    if (total % 3 != 0){
      return false;
    }

    int leftIndex = 1;
    int rightIndex = arr.size() - 2;
    int average = total / 3;
    while (leftIndex < rightIndex){
      if (leftIndex < rightIndex && left != average){
        left += arr[leftIndex];
        leftIndex += 1;
      }
      if (leftIndex < rightIndex && right != average){
        right += arr[rightIndex];
        rightIndex -= 1;
      }
      if (left == average && right == average){
        return true;
      }
    }
    return false;
  }
};

File: 1013. Partition Array Into Three Parts With Equal Sum/solution.py
class Solution(object):
  def canThreePartsEqualSum(self, arr):
    """
    :type arr: List[int]
    :rtype: bool
    """
    left = arr[0]
    right = arr[-1]
    total = 0
    for num in arr:
      total += num
    
    if total % 3 != 0:
      return False
    
    leftIndex = 1
    rightIndex = len(arr) - 2
    average = total / 3
    while leftIndex < rightIndex:
      if leftIndex < rightIndex and left != average:
        left += arr[leftIndex]
        leftIndex += 1
      if leftIndex < rightIndex and right != average:
        right += arr[rightIndex]
        rightIndex -= 1
      if left == average and right == average:
        return True
    return False